### PR TITLE
feature/cplp-2235-conformity-document-type-app-api

### PR DESCRIPTION
**Pull request by qxz2mqe**
https://jira.catena-x.net/browse/CPLP-2235

### CODE REVIEW

File: src/marketplace/Offers.Library/Service/OfferService.cs [hunk 1]
```diff
@@ -449,19 +449,42 @@ public void CreateOrUpdateOfferLicense(Guid offerId, string licenseText, (Guid O
...
     }
+
+    /// <inheritdoc/>
+    public async Task SubmitOfferAsync(Guid offerId, string iamUserId, OfferTypeId offerTypeId, IEnumerable<NotificationTypeId> notificationTypeIds, IDictionary<string,IEnumerable<string>> catenaAdminRoles, IEnumerable<DocumentTypeId> submitAppDocumentTypeIds)
+    {
+        var offerDetails = await GetOfferReleaseData(offerId, offerTypeId);
+        var invalidDocumentType = offerDetails.DocumentTypeIds.Except(submitAppDocumentTypeIds);
+        if (invalidDocumentType.Any())
+        {
+            throw new ConflictException($"invalid DocumenType: [{String.Join(", ", invalidDocumentType)}]");
```
Comment (Phil91) [inline src/marketplace/Offers.Library/Service/OfferService.cs:460]: The implementation currently deviates from the described behavior in the task.

At the moment it is checked if other DocumentTypeIds are uploaded as well, the check has to be adapted.

It should be checked whether specifically all of the three "Conformity_Approval_Business_Apps", "APP_LEADIMAGE", "APP_IMAGE" have been uploaded. If this is not the case, different error messages should be returned.

**Review by Phil91 — CHANGES_REQUESTED** [review]
Please add a PR for the config changes in the CD repository as well

File: src/marketplace/Offers.Library/Service/OfferService.cs (hunk 1 above)
Comment (qxz2mqe) [reply src/marketplace/Offers.Library/Service/OfferService.cs]: fixed

File: src/marketplace/Offers.Library/Service/OfferService.cs [hunk 2]
```diff
@@ -331,10 +331,10 @@ public void CreateOrUpdateOfferLicense(Guid offerId, string licenseText, (Guid O
     public async Task SubmitOfferAsync(Guid offerId, string iamUserId, OfferTypeId offerTypeId, IEnumerable<NotificationTypeId> notificationTypeIds, IDictionary<string,IEnumerable<string>> catenaAdminRoles, IEnumerable<DocumentTypeId> submitAppDocumentTypeIds)
     {
         var offerDetails = await GetOfferReleaseData(offerId, offerTypeId);
-        var invalidDocumentType = offerDetails.DocumentTypeIds.Except(submitAppDocumentTypeIds);
-        if (invalidDocumentType.Any())
+        var isvalidDocumentType = offerDetails.DocumentTypeIds.All(x=>submitAppDocumentTypeIds.Contains(x));
```
Comment (Phil91) [inline src/marketplace/Offers.Library/Service/OfferService.cs:334]: please change to `submitAppDocumentTypeIds.All(x=> offerDetails.DocumentTypeIds.Contains(x))`

File: src/marketplace/Offers.Library/Service/OfferService.cs [hunk 3]
```diff
@@ -331,10 +331,10 @@ public void CreateOrUpdateOfferLicense(Guid offerId, string licenseText, (Guid O
...
+        if (!isvalidDocumentType)
         {
-            throw new ConflictException($"invalid DocumenType: [{String.Join(", ", invalidDocumentType)}]");
+            throw new ConflictException($"Conformity_Approval_Business_Apps,  APP_LEADIMAGE, APP_IMAGE are mandatory document type");
```
Comment (Phil91) [inline src/marketplace/Offers.Library/Service/OfferService.cs:337]: Please change to `$"{string.Join(",", submitAppDocumentTypeIds)} are mandatory document types`

File: tests/marketplace/Offers.Library.Tests/Service/OfferServiceTests.cs [hunk 4]
```diff
@@ -650,7 +650,7 @@ public async Task SubmitOffer_WithInvalidDocumentType_ThrowsConflictException()
 
         // Assert
         var result = await Assert.ThrowsAsync<ConflictException>(Act).ConfigureAwait(false);
-        result.Message.Should().StartWith($"invalid DocumenType: [{String.Join(", ", DocumentTypeId.SELF_DESCRIPTION)}]");
+        result.Message.Should().StartWith($"Conformity_Approval_Business_Apps,  APP_LEADIMAGE, APP_IMAGE are mandatory document type");
```
Comment (Phil91) [inline tests/marketplace/Offers.Library.Tests/Service/OfferServiceTests.cs:653]: Please adjust the error message to only contain the values which are provided within the parameters

**Review by Phil91 — CHANGES_REQUESTED** [review]

File: src/marketplace/Offers.Library/Service/OfferService.cs (hunk 3 above)
Comment (qxz2mqe) [reply src/marketplace/Offers.Library/Service/OfferService.cs]: Fixed

File: tests/marketplace/Offers.Library.Tests/Service/OfferServiceTests.cs (hunk 4 above)
Comment (qxz2mqe) [reply tests/marketplace/Offers.Library.Tests/Service/OfferServiceTests.cs]: fixed

File: src/marketplace/Offers.Library/Service/OfferService.cs (hunk 1 above)
Comment (qxz2mqe) [reply src/marketplace/Offers.Library/Service/OfferService.cs]: fixed

File: src/marketplace/Offers.Library/Service/OfferService.cs (hunk 2 above)
Comment (qxz2mqe) [reply src/marketplace/Offers.Library/Service/OfferService.cs:334]: I tried to implement in this way as you suggested here but it is giving wrong result. Please verify it. I tested in postman and the way i implemented giving expected output

Comment (Phil91) [reply src/marketplace/Offers.Library/Service/OfferService.cs:334]: just checked it. for example add DocumentTypeId.APP_CONTRACT in line 671 in the OfferServiceTests, the test will fail with the current implementation.

But it is allowed that more than the documentTypeIds of the settings are uploaded. But those configured must be uploaded

**Review by Phil91 — CHANGES_REQUESTED** [review]

**Review by Phil91 — APPROVED** [review]

**Pull request merged**